reference/ast-examples.md

```diff
@@ -0,0 +1,147 @@
+# AST Element Examples
+
+When building recipes, it's important to understand how certain elements in the OpenRewrite Abstract Syntax Tree (AST) correspond to code. You couldn't, for example, properly rename a variable with a recipe unless you knew that `J.Identifier` is how variable names are represented in the AST.
+
+To help assist you with identifying these elements, this guide will provide a sample chunk of code, highlight which AST elements correspond to which piece(s) of code, and explain a bit about each element. At the end, you'll also find instructions for how to use the debugger to learn more about the elements yourself.
+
+* [Code Sample](#code)
+  * [Binary](#binary)
+  * [Block](#block)
+  * [ClassDeclaration](#classdeclaration)
+  * [CompilationUnit](#compilationunit)
+  * [Expression](#expression)
+  * [FieldAccess](#fieldaccess)
+  * [Identifier](#identifier)
+  * [MethodDeclaration](#methoddeclaration)
+  * [MethodInvocation](#methodinvocation)
+  * [NewClass](#newclass)
+  * [Statement](#statement)
+  * [VariableDeclarations](#variabledeclarations)
+* [Using the Debugger To Detect Elements](#using-the-debugger-to-detect-elements)
+
+## Code
+
+Below is a simple Java class whose entire purpose is to demonstrate different AST elements. Each of the following sections will highlight different parts of this code to demonstrate which chunks correspond to which AST element.
+
+```java
+package org.openrewrite;
+                
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@AnAnnotation
+public class A {
+
+    List<Integer> a = new ArrayList<>();
+
+    int foo() {
+        int a = 1 + 2, b = 3, c;
+        this.a = this.a.stream()
+                .map(it -> it + 1)
+                .collect(Collectors.toList());
+        return a;
+    }
+}
+
+@interface AnAnnotation {}
+```
+
+### Binary
+
+A [Binary element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L597-L705) is any operation with a left and right side (usually separated by something like a `+`).
+
+![Binary Element Example](<../.gitbook/assets/Binary.png>)
+
+### Block 
+
+A [Block element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L712-L851) is the code contained between curly braces. Blocks can be nested inside of each other.
+
+![Block Element Example](<../.gitbook/assets/Block.png>)
+
+### ClassDeclaration
+
+A [ClassDeclaration element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L1062-L1336) contains all of the code for any Java class. 
```

Comment:
I think it would be a good idea to remind users that a ClassDeclaration can be nested inside another class: 

```
public class A {
...

   private class B {
   ...
   }
}
```